### ISSUE TITLE
InSingletonScope per key or InScope transient key

### ISSUE THREAD

**Issue by brenovieira** [ninject/Ninject#195]
I'm building a [multi tenant](https://en.wikipedia.org/wiki/Multitenancy) web api.

In my `NinjectModule`, I have several singletons per `tenantId` which is passed in each request (encrypted).

To achieve this, I'm creating a static dictionary for each class which holds these class objects per key (`tenantId`).
Something like:
```
public class Bar { } // I need just one of this per key

public class BarProvider
{
   private readonly ConcurrentDictionary<string, Bar> _cache =
      new ConcurrentDictionary<string, Bar>();

   public Bar GetBar(string key)
   {
      if (!_cache.ContainsKey(key)) _cache[key] = new Bar();
      
      return _cache[key];
   }
}

public class NinjectConfig : NinjectModule
{
   Bind<BarProvider>().ToSelf().InSingletonScope();
   Bind<Bar>()
      .ToMethod(ctx  =>
      {
         var key = HttpContext.Current.Request.Headers["key"];
         return ctx.Kernel.Get<BarProvider>().GetBar(key);
      })
      .InTransientScope();
}
```

But after creating some of these "Providers", I'm trying to discover a way of Ninject do this for me.
I tried to use `InScope`, like this:
```
Bind<Bar>()
   .ToSelf()
   .InScope(ctx  => HttpContext.Current.Request.Headers["key"]);
```

What I was trying to do here is to tell `Ninject` to create a `Bar` per `key`, but as `key` is a transient object (just a value without a permanent reference), it is diposed, collected by GC and it doesn't work as I expected.

I also looked up [Ninject.Extensions.Factory](https://github.com/ninject/Ninject.Extensions.Factory), but I think it doesn't solve my needs as well.

Is there another way in Ninject to do this ?
If not, what do you about creating a `InSingletonScope(string key)` or `InScope()` that works with values ?

Thanks

**Comment by PehrGit** [comment]
Did you ever figure this out?

**Comment by brenovieira** [comment]
@PehrGit , no.

I just changed my class `BarProvider` to use generics, something like `BarProvider<TKey, TValue>` which has a concurrentDictionary and Get and Create methods.

and in NinjectConfig I'm using like this:
```
var provider = new BarProvider<MyKey, MyValue>();
Bind<IMyValue>().ToMethod(ctx => barProvider.Get(ctx));
```

I would like to Ninject do this for me, but I noticed last release of Ninject was long time ago and I think the maintainers could take too much time to release another version.

So I did it myself and closed the issue.
